cpp/include/cudf/hashing.hpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+namespace cudf {
+
+enum class hash_func {
+  HASH_MURMUR3=0, ///< Murmur3 hash function
+  HASH_IDENTITY,  ///< Identity hash function that simply returns the key to be hashed
+};
+
+/** --------------------------------------------------------------------------*
+ * @brief Computes the hash values of the rows in the specified columns of the 
+ * input columns and bins the hash values into the desired number of partitions. 
+ * Rearranges the input columns such that rows with hash values in the same bin 
+ * are contiguous.
+ * 
+ * @param[in] input The input set of columns
+ * @param[in] columns_to_hash Indices of the columns in the input set to hash
+ * @param[in] num_partitions The number of partitions to rearrange the input rows into
+ * @param[in] hash The hash function to use
+ * 
+ * @returns A pair of <partioned output, partition offsets>, where the partitioned
+ * output is a new table with rows from the input reordered into partitions and
+ * the partion offsets are a vector of offsets to the first index of each partition
+ * ----------------------------------------------------------------------------**/
+std::pair<std::unique_ptr<table>, std::vector<size_type>>
+hash_partition(table_view const& input,
```

Comment:
This should probably be updated to return `std::vector<std::unique_ptr<table>>`, i.e., it actually does the partitioning of `input` into discrete tables.

Reply:
Good point, that's a lot cleaner than the old signature